rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+* Exposing this new behavior into the docker container, and the library work necessary to make this feature usable and natural in other languages.
+* Adding atomic functionality to the datastore API and internals to make distributed mutation (not just key replacement) possible
+
+### Cluster Coordination Work
+
+The new endpoints have some extra logic that they need to tackle, as this is the first time the nodes of the cluster are actually communicating with each other. The LogRendezvousHash module needs to be updated on the *actual* state of the cluster versus the expected state defined in the DNS records: which nodes are actually up or down from its perspective? That determines which nodes are considered for ownership of the datastore data. RendezvousHash (and our own LogRendezvousHash) are stable, in that the top `n` nodes after one of them goes down will still contain `n - 1` of the original set, so not too much data thrashing occurs when a node goes down or comes online.
+
+Further, while a full network partition between two groups of nodes will result in a "split brain" scenario in an unavoidable fashion, with LogRendezvousHash it is possible to counteract a partial split by way of a request forwarding mechanism. If a node cannot directly reach another node, it should try to ask all other nodes in its list if any of them can reach the node. If so, the first one to respond in the positive becomes the new "router" that all requests to that node are forwarded through, with periodic attempts in the background to remove the extra hop by re-polling. Only if all nodes a node can reach report that they also cannot reach the node in question should it be marked as "down" and assignment of keys shuffled with the new, reduced set. This is fundamentally how the internet routes packets to unknown MAC addresses, and can guard against misconfigured firewalls/routers black hole-ing packets that they shouldn't be, but only works when there is an unaffected third party, so requires at least three separate sub-clusters to work.
+
+For example, let's presume a cluster has been deployed to all three major cloud providers: AWS, Azure, and GCP. Now let's presume that GCP has a network configuration error that makes AWS nodes unreachable to it and vice versa. When the GCP nodes want to get data that was previously assigned to AWS, they will fail to contact the AWS node, then try to ask all other nodes (which would include the other AWS nodes) which ones can reach the AWS node in question. The GCP nodes would report a failure, but the Azure nodes would report a success, and one of those nodes will be chosen effectively at random to be the node that will route to the AWS node. This process would repeat across every GCP node and AWS node combination with an effectively random assignment of Azure nodes to handle the load, which should lead to an approximately even level of traffic distribution between the Azure nodes.
```

Comment:
Do three separate sub-clusters necessarily mean three sub clusters across regions or cloud providers, or simply if there is one unreachable/failing node within a 3-node cluster in the same region and cloud provider? That seems like a more likely scenario that we should also cover where one node receives a request that blows up the process.

Reply:
One unreachable/failed node is automatically already dealt with by the RendezvousHash logic as described back in the 90s, so I didn't consider it worth going over in this RFC because it's more trivial.

Reply:
It is worthwhile to point that out even if it is trivial since the reader of the RFC might not be familiar? I understand it makes the RFC longer and that was also probably a factor. I would have though about the cluster coordination and mutable distributed global state as separate RFCs, but I also understand why you made them into the same RFC after reading through it.

Reply:
Sure. I can add that, but I won't remove this example because it was the more important piece to figure out the high-level algorithm. (Just using RendezvousHash by itself here didn't solve the partial partition problem.)